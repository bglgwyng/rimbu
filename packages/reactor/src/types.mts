import type { Actor } from '@rimbu/actor';
import type { Deep } from '@rimbu/deep';

export type SelectorEntry = {
  value: unknown;
  listeners: Map<Actor.Listener, { count: number }>;
};

export type SelectorCache = Map<Deep.Selector.Shape<any>, SelectorEntry>;
