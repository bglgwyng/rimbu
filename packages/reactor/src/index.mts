export { Reactor, useForceRerender } from './internal.mjs';
