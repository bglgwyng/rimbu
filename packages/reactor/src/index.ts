export { Reactor, useForceRerender } from './internal';
