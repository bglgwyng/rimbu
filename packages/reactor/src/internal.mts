// no dependencies
export * from './types.mjs';
export * from './use-force-rerender.mjs';
export * from './update-selector-value.mjs';
export * from './register-selector.mjs';
export * from './unregister-selector.mjs';

// internal dependencies
export * from './reactor.mjs';
