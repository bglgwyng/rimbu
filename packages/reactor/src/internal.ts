export * from './controls/ObsStream';

export * from './hooks/useBindings';
export * from './hooks/useConst';
export * from './hooks/useCreateObs';
export * from './hooks/useActor';
export * from './hooks/useActorState';
export * from './hooks/useActorSubscribe';
export * from './hooks/useSubscribeUpdateUI';
export * from './hooks/useUpdateUI';
