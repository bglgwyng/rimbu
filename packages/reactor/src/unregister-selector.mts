import type { Actor } from '@rimbu/actor';
import type { Deep } from '@rimbu/deep';

import type { SelectorCache } from './internal';

export function unregisterSelector(
  selector: Deep.Selector.Shape<any>,
  listener: Actor.Listener,
  selectorCache: SelectorCache
): void {
  const updateEntry = selectorCache.get(selector);

  if (undefined === updateEntry) {
    return;
  }

  const counter = updateEntry.listeners.get(listener);

  if (undefined === counter) {
    return;
  }

  if (counter.count === 1) {
    updateEntry.listeners.delete(listener);

    if (updateEntry.listeners.size === 0) {
      selectorCache.delete(selector);
    }
  } else {
    counter.count--;
  }

  if (Array.isArray(selector)) {
    for (const subSelector of selector) {
      unregisterSelector(subSelector, listener, selectorCache);
    }
  } else if (typeof selector === 'object') {
    for (const key in selector) {
      unregisterSelector(selector[key], listener, selectorCache);
    }
  }
}
