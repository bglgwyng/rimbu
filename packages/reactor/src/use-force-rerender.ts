import React from 'react';

export function useForceRerender(): () => void {
  const setState = React.useState(0)[1];

  return (): void => setState((v) => (v + 1) | 0);
}
