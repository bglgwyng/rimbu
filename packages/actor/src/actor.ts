import { Obs } from './internal';

/**
 * Am `Actor` that has an `Obs` instance so that its state can be observed and modified. Its main aim is to allow
 * custom methods to be added to an `Obs`.
 * @typeparam T - the state type
 */
export interface Actor<T, D = unknown> extends Actor.Readonly<T & D> {
  /**
   * Returns the `Obs` instance of this `Actor`.
   */
  readonly obs: Obs<T, D>;
}

export namespace Actor {
  /**
   * A readonly `Actor` that has a readonly `Obs` instance so that its state can be observed. Its main aim is to allow
   * custom methods to be added to an `Obs.Readonly`.
   * @typeparam T - the state type
   */
  export interface Readonly<T> {
    /**
     * Returns the readonly `Obs` instance of this `Actor`.
     */
    readonly obsReadonly: Obs.Readonly<T>;
    /**
     * Returns the state of the associated `Obs` instance.
     */
    readonly state: this['obsReadonly']['state'];
  }

  export type StateType<A extends Actor.Readonly<unknown>> =
    A extends Actor.Readonly<infer S> ? S : never;

  class Impl<T, P extends Record<string, unknown>, D> implements Actor<T, D> {
    constructor(readonly actor: Actor<T, D>, props: P) {
      Object.assign(this, props);
    }

    get obs() {
      return this.actor.obs;
    }

    get obsReadonly() {
      return this.actor.obsReadonly;
    }

    get state() {
      return this.actor.state;
    }
  }

  /**
   * Returns an `Actor` instance based on the given input `Actor` (which may also be an `Obs`), and adds the methods and properties of the
   * given `props` object to it.
   * @param actor - the Actor or Obs instance to add the given `props` to
   * @param props - the properties to add to the `actor`.
   * @example
   * const o = Obs.create({ count: 0 })
   * const actor = Actor.from(o, {
   *   inc(amount: number) {
   *     o.patchState({ count: v => v + amount })
   *   }
   *   reset() {
   *     o.setState({ count: 0 })
   *   }
   * })
   *
   * console.log(actor.state)
   * // => { count: 0 }
   * actor.inc(2)
   * console.log(actor.state)
   * // => { count: 2 }
   * actor.reset()
   * console.log(actor.state)
   * // => { count: 0 }
   */
  export const from: {
    <T, P extends Record<string, unknown>, D = unknown>(
      actor: Actor<T, D>,
      props: P
    ): Actor<T, D> & P;
    <T, P extends Record<string, unknown>>(
      actor: Actor.Readonly<T>,
      props: P
    ): Actor.Readonly<T> & P;
  } = <T, P extends Record<string, unknown>>(
    actor: Actor.Readonly<T>,
    props: P
  ) => {
    return new Impl(actor as any, props);
  };
}
