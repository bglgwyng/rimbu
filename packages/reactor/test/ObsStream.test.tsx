import { Obs } from '@rimbu/actor';
import { List } from '@rimbu/core';
import { act, render } from '@testing-library/react';
import React from 'react';
import { ObsStream } from '../src';

describe('ObsStream', () => {
  it('renders empty list', () => {
    const source = Obs.create([] as number[]);

    const Component = (props: { item: number }) => <div>{props.item}</div>;

    const { container } = render(
      <ObsStream
        source={() => source}
        toKey={String}
        toProps={(item) => ({ item })}
        component={Component}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('renders non-empty list', () => {
    const source = Obs.create([1, 2, 3] as number[]);
    const Component = (props: { item: number }) => <div>{props.item}</div>;

    const { container } = render(
      <ObsStream
        source={() => source}
        toKey={String}
        toProps={(item) => ({ item })}
        component={Component}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('creates only elements for newly added values', () => {
    const Component = jest
      .fn()
      .mockImplementation((props: { item: number }) => <div>{props.item}</div>);

    const itemObs = Obs.create({
      items: List.of(1, 2, 3),
      value: 'a',
    });

    const Control = React.memo(() => {
      return (
        <ObsStream
          source={() => itemObs.mapReadonly((s) => s.items)}
          component={Component}
          toKey={String}
          toProps={(item) => ({ item })}
        />
      );
    });

    const { container } = render(<Control />);

    expect(container.children.length).toBe(3);

    expect(Component).toBeCalledTimes(3);
    Component.mockClear();

    act(() => itemObs.patchState({ items: (v) => v.concat([4, 5]) }));

    expect(container.children.length).toBe(5);
    expect(Component).toBeCalledTimes(2);
  });

  it('does rerender when observable value changes', () => {
    const Component = jest
      .fn()
      .mockImplementation((props: { item: number }) => <div>{props.item}</div>);

    const itemObs = Obs.create({ items: List.of(1, 2, 3), value: 'a' });

    const Control = React.memo(() => {
      return (
        <ObsStream
          source={() => itemObs.mapReadonly((s) => s.items)}
          component={Component}
          toKey={String}
          toProps={(item) => ({ item })}
        />
      );
    });

    const { container } = render(<Control />);

    expect(container.children.length).toBe(3);

    expect(Component).toBeCalledTimes(3);
    Component.mockClear();

    act(() =>
      itemObs.patchState({ items: (v) => v.concat([4, 5]), value: 'b' })
    );

    expect(container.children.length).toBe(5);
    expect(Component).toBeCalledTimes(2);
  });

  it('removes items', () => {
    const Component = jest
      .fn()
      .mockImplementation((props: { item: number }) => <div>{props.item}</div>);

    const itemObs = Obs.create({
      items: List.of(1, 2, 3).asNormal(),
    });

    const Control = React.memo(() => {
      return (
        <ObsStream
          source={() => itemObs.mapReadonly((s) => s.items)}
          component={Component}
          toKey={String}
          toProps={(item) => ({ item })}
        />
      );
    });

    const { container } = render(<Control />);

    expect(container.children.length).toBe(3);

    expect(Component).toBeCalledTimes(3);
    Component.mockClear();

    act(() => itemObs.patchState({ items: (v) => v.drop(1) }));

    expect(container.children.length).toBe(2);
    expect(Component).not.toBeCalled();
  });
});
