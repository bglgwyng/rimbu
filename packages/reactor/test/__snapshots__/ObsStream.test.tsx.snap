// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObsStream renders empty list 1`] = `<div />`;

exports[`ObsStream renders non-empty list 1`] = `
<div>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;
